src/box/txn_limbo.c

```diff
@@ -457,6 +457,13 @@ txn_limbo_read_confirm(struct txn_limbo *limbo, int64_t lsn)
 		assert(e->txn->signature >= 0);
 		txn_complete_success(e->txn);
 	}
+	/*
+	 * Track CONFIRM lsn on replica in order to detect split-brain by
+	 * comparing existing confirm_lsn with the one arriving from a remote
+	 * instance.
+	 */
+	if (limbo->confirmed_lsn < lsn)
+		limbo->confirmed_lsn = lsn;
```

Comment:
It can't on a leader, but it can on a replica.
Since both replica and leader issue `txn_limbo_read_confirm`, it was easier for me to fix the replica this way.
In other words, this branch only works on everyone but the synchro queue owner.

Having explained this to you, I decided to refactor it this way:
```diff
diff --git a/src/box/txn_limbo.c b/src/box/txn_limbo.c
index b9c0cf85c..dadabfc16 100644
--- a/src/box/txn_limbo.c
+++ b/src/box/txn_limbo.c
@@ -469,8 +469,10 @@ txn_limbo_read_confirm(struct txn_limbo *limbo, int64_t lsn)
         * comparing existing confirm_lsn with the one arriving from a remote
         * instance.
         */
-       if (limbo->confirmed_lsn < lsn)
+       if (limbo->owner_id != instance_id) {
+               assert(limbo->confirmed_lsn <= lsn);
                limbo->confirmed_lsn = lsn;
+       }
 }
 
 /**
```

With this change some replication tests on commit "txn_limbo: track CONFIRM lsn on replicas" fail with assertion due to these lines:
https://github.com/tarantool/tarantool/blob/b3f462bf442a98dd2ae61b6cd7f863d2e2e8c1e8/src/box/txn_limbo.c#L867-L883

They are obsolete in presence of split-brain detection, and everything works fine and passes once the commit "txn_limbo: filter incoming synchro requests" is applied

Reply:
Uh oh, this doesn't work, unfortunately. You also have to track confirmed_lsn while recovering an instance.

I tried the following diff, but it still fails the assertion occasionally.

```diff
diff --git a/src/box/txn_limbo.c b/src/box/txn_limbo.c
index dadabfc16..f3c9bb1d8 100644
--- a/src/box/txn_limbo.c
+++ b/src/box/txn_limbo.c
@@ -465,11 +465,11 @@ txn_limbo_read_confirm(struct txn_limbo *limbo, int64_t lsn)
                txn_complete_success(e->txn);
        }
        /*
-        * Track CONFIRM lsn on replica in order to detect split-brain by
-        * comparing existing confirm_lsn with the one arriving from a remote
-        * instance.
+        * Track CONFIRM lsn on replica and on recovering queue owner in order
+        * to detect split-brain by comparing existing confirm_lsn with the
+        * one arriving from a remote instance.
         */
-       if (limbo->owner_id != instance_id) {
+       if (limbo->owner_id != instance_id || !box_is_configured()) {
                assert(limbo->confirmed_lsn <= lsn);
                limbo->confirmed_lsn = lsn;
        }
```

I don't have any ideas on how to fix this now, so returned to the original version.

Reply:
Hm, but how can it be less than the last received one on the replicas?

Reply:
Oh. Now I get your question.
When running this code (`if confirmed_lsn < lsn`) on replica, it's only possible for `confirmed_lsn` to be equal to `lsn` while reading a PROMOTE or DEMOTE. `txn_limbo_req_prepare` sets limbo->confirmed_lsn to request lsn, and then we enter `txn_limbo_req_commit`, `txn_limbo_read_promote`. The latter in turn calls `txn_limbo_read_confirm`, where `confirmed_lsn == lsn`.

 It's never possible that `confirmed_lsn > lsn` on a replica. Unless we receive a broken packet.

Now, as I understand, there are some cases when `confirmed_lsn > lsn` when recovering a limbo owner.
I tried to avoid the assertions by adding `if owner_id != 0 && owner_id != instance_id` but that didn't help either.

This is where I stopped investigating and returned to the original simple version with
`if confirmed_lsn < lsn then confirmed_lsn = lsn`.

So, short answer is: lsn is always >= confirmed_lsn on a replica, but it seems the opposite might happen on master during recovery.